### ISSUE TITLE
UPnPAction read xmlElement function

### ISSUE THREAD

**Issue by meisbedi** [bjtj/swift-upnp-tools#9]
if the UPnPAction read function hits an element, that does not have the name "argumentList" and does not contain further elements  (element.elements == nil) the current implementation:

guard let elements = element.elements!.isEmpty) 

creates a runtime error (unexpectedly found nil).

 public static func read(xmlElement: XmlElement) -> UPnPAction? {
        let action = UPnPAction()
        guard let elements = xmlElement.elements else {
            print("UPnPAction::read() error - no xml element")
            return nil
        }
        for element in elements {
            guard let name = element.name else {
                print("UPnPAction::read() error - no element name found")
                continue
            }
            
            if name == "argumentList" {
                guard let argListElements = element.elements else {
                    print("UPnPAction::read() error - wrong arg list")
                    continue
                }
                for argElement in argListElements {
                    guard let argument = UPnPActionArgument.read(xmlElement: argElement) else {
                        print("UPnPAction::read() error - parse argument failed")
                        continue
                    }
                    action.arguments.append(argument)
                }
            } else {
                **_guard let elements = element.elements, elements.isEmpty else_** {
                    continue
                }
                guard let name = element.name else {
                    continue
                }
                guard let value = element.firstText!.text else {
                    continue
                }
                action[name] = value
            }
        }
        return action
    }

**Comment by bjtj** [comment]
I fixed in tag 0.1.21

Thank you!

**Comment by meisbedi** [comment]
Hi there,

today I submitted and issue on the swift-xml package and I wanted to take the opportunity and give you some general feedback on your swift-upnp-tools package.

I have used your package in my app development for iOS since late last year. I am very grateful for your support to address the issues that came up during my development and testing. I have not experienced any crashes during my testing since the beginning of the year. And the general function seems to be robust. I’m very happy, that I decided to give your package a try.

The only issues I’m still experiencing are race conditions due to the asynchronous structure of the operation. I have addressed step by step most of these issues by introducing semaphores, but this might be a rather crude approach. But I’m not a SW professional and it does the job. Most of the problems occur during the suspend/resume cycle of the app. But so far, it has never affected the function of the package.

I’m currently working on the release candidate of my app and I might take the time and review the whole situation once more and try to clean up the mess I have created. If I succeed, I will let you know about the approach I took to make swift-upnp-tools and swift-http-server thread save.

Best Regards
Dieter
